diff --git a/dlls/ntdll/file.c b/dlls/ntdll/file.c
index 46fea51..216a6a7 100644
--- a/dlls/ntdll/file.c
+++ b/dlls/ntdll/file.c
@@ -1532,6 +1532,10 @@ static NTSTATUS server_ioctl_file( HANDLE handle, HANDLE event,
         req->async.event    = wine_server_obj_handle( event );
         req->async.cvalue   = cvalue;
         wine_server_add_data( req, in_buffer, in_size );
+        if ((code & 3) == METHOD_IN_DIRECT)
+        {
+            wine_server_add_data( req, out_buffer, out_size );
+        }
         wine_server_set_reply( req, out_buffer, out_size );
         status = wine_server_call( req );
         wait_handle = wine_server_ptr_handle( reply->wait );
diff --git a/dlls/ntoskrnl.exe/ntoskrnl.c b/dlls/ntoskrnl.exe/ntoskrnl.c
index 0e87b82..e5a2c91 100644
--- a/dlls/ntoskrnl.exe/ntoskrnl.c
+++ b/dlls/ntoskrnl.exe/ntoskrnl.c
@@ -364,7 +364,11 @@ static NTSTATUS dispatch_ioctl( const irp_params_t *params, void *in_buff, ULONG
         {
             memcpy( out_buff, in_buff, in_size );
             in_buff = out_buff;
-        }
+        } else if ((params->ioctl.code & 3) == METHOD_IN_DIRECT)
+        {
+            in_size -= out_size;
+            memcpy( out_buff, (PBYTE) in_buff + in_size, out_size );
+         }
     }
 
     /* note: we abuse UserIosb to store the server handle to the ioctl */
